### ISSUE TITLE
Performing a gradual migration from Prometheus+Thanos to Victoriametrics 

### ISSUE THREAD

**Issue by hartfordfive** [VictoriaMetrics/VictoriaMetrics#6593]
### Is your question request related to a specific component?

Victoriametrics, vminsert, vmagent 

### Describe the question in detail

I would like to know if it's possible to have both Prometheus remote write metrics to vminsert and vmagent scrape the same metrics at the same time while ensuring all incoming metrics are properly deduplicated?   I would be interested in this type of configuration of possible as it could allow us to gradually migrate and decommission individual Prometheus shards (HA pair) by simply turning off the remote write on one Prometheus shard at a time.  This also allows us to quickly re-enable remote-write if we notice an issue (such as missing metrics) on the Victoriametrics cluster.   

I would assume this may be possible and would be related to the deduplication functionality of vminsert?  Also, would the scrape interval of the Prometheus shards be required to be aligned with the scrape interval of vmagent?  I would appreciate some clarification and any additional relevant details.

Thank you for your help.

### Troubleshooting docs

- [X] General - https://docs.victoriametrics.com/troubleshooting/
- [X] vmagent - https://docs.victoriametrics.com/vmagent/#troubleshooting
- [ ] vmalert - https://docs.victoriametrics.com/vmalert/#troubleshooting

**Comment by hagen1778** [comment]
Hello @hartfordfive!

> I would assume this may be possible and would be related to the deduplication functionality of vminsert?

Deduplication functionality belongs to vmstorage and vmselect components - see https://docs.victoriametrics.com/#deduplication

> Also, would the scrape interval of the Prometheus shards be required to be aligned with the scrape interval of vmagent?

Deduplication will keep one sample per series, per interval. So if source A wrote two samples of one time series at 15s and 29s, and source B wrote one sample of the same time series at 27s - then with `dedup.minScrapeInterval=30s` only sample from `29s` will be kept, the rest discarded.
In short, yes, the scrape intervals need to be the same for deduplication to always keep the sample from the same replica (the replica that writes the last sample on the interval).

Please note, for deduplication to work all sources need to work identical series (all labels need to match).

**Comment by hartfordfive** [comment]
When you say all labels need to match, does this also include labels specified via the external_labels configuration?

**Comment by hagen1778** [comment]
@hartfordfive wdyt about alternative approach for gradual migration: splitting Prometheus and vmagent data by tenants?
So Prometheus will write the data to tenant 0, and vmagent will write to tenant 1.

Tenants are completely isolated, there will be no interference between data. This will give you an option to easily compare query results between tenant 0 and tenant 1, without thinking about deduplication complexity or scrape intervals. You can have separate datasources in Grafana, separate rulers, etc. Once you're happy with the results, you can start sunsetting Prometheus/Thanos, knowing that tenant 1 already has all the data and that data is correct.

**Comment by hartfordfive** [comment]
I actually didn't even think about this solution. Thanks for suggesting it.  I  think this could definitely do the job although could we "force" delete all the data for the first tenant once we have considered the migration successful?

**Comment by hagen1778** [comment]
> I think this could definitely do the job although could we "force" delete all the data for the first tenant once we have considered the migration successful?

It can be deleted, yes. You also can set [custom retention period per tenant](https://docs.victoriametrics.com/#retention-filters). Please note, currently VictoriaMetrics doesn't support [cross-tenant read queries](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/6346).

**Comment by jiekun** [comment]
I'm closing this as answered. Please reopen it if you have more question on this topic.
